File: java/Stack/CarFleet.java
package java.Stack;

import java.util.Arrays;
import java.util.Stack;

public class CarFleet {

    // Helper class to store car data
    private static class Car {
        int position;
        int speed;
        
        Car(int position, int speed) {
            this.position = position;
            this.speed = speed;
        }
    }

    public int carFleet(int target, int[] position, int[] speed) {

        int n = position.length;
        
        // Step 1: Create array of car data and sort by position (descending)
        Car[] cars = new Car[n];
        for (int i = 0; i < n; i++) {
            cars[i] = new Car(position[i], speed[i]);
        }
        
        // Sort by position in descending order (closest to target first)
        Arrays.sort(cars, (a, b) -> Integer.compare(b.position, a.position));
        
        // Step 2: Use monotonic stack to track fleets
        Stack<Double> stack = new Stack<>();
        
        for (Car car : cars) {
            // Calculate time needed to reach target
            double time = (double)(target - car.position) / car.speed;
            
            // If stack is empty or current car takes longer than the fleet ahead,
            // it forms a new fleet
            if (stack.isEmpty() || time > stack.peek()) {
                stack.push(time);
            }
            // Else, current car joins the fleet ahead (faster car gets slowed down)
        }
        
        return stack.size();
        
    }
}

File: python/Stack/CarFleet.py
"""
853. Car Fleet
Difficulty: Medium
Topics: Array, Stack, Monotonic Stack, Sorting
"""

def carFleet(target: int, position: list[int], speed: list[int]) -> int:
    # Step 1: Pair position and speed, then sort by position (closest to target first)
    cars = sorted(zip(position, speed), reverse=True)

    # Step 2: Calculate time to reach target for each car
    times = [(target-pos)/spd for pos, spd in cars]
    # Step 3: Use monotonic stack to count fleets
    stack = []
    
    for time in times:
        # If current car arrives later than the car in front, it forms a new fleet
        if not stack or time > stack[-1]:
            stack.append(time)
        # Else, current car joins the fleet ahead (slower car dominates)
        # We don't need to push because the fleet speed is determined by the slower car
    
    return len(stack)